### PR TITLE
tests: Rework callback expectation checks with `PlotRefreshTester`

### PR DESCRIPTION
**Pull request by xdustinface**

### CODE REVIEW

File: tests/plotting/test_plot_manager.py [hunk 1]
```diff
@@ -76,10 +76,64 @@ def drop(self, path: Path):
...
 
+class PlotRefreshTester:
+    plot_manager: PlotManager
+    expected_result: PlotRefreshResult
+    expected_result_matched: bool
+
+    def __init__(self, root_path: Path):
+        self.plot_manager = PlotManager(root_path, self.refresh_callback)
+        self.expected_result = PlotRefreshResult()
+        self.expected_result_matched = False
```
Comment (arvidn) [inline tests/plotting/test_plot_manager.py:87]: the assignments to `self.expected_result_matched` and `expected_result` are "dead". they will always be overwritten in `run()`.

It doesn't really seem like these belong as members in this class, but this class also seems to be a round-about way to include these in the closure of the callback.

File: tests/plotting/test_plot_manager.py [hunk 2]
```diff
@@ -76,10 +76,64 @@ def drop(self, path: Path):
...
+                        return
+                else:
+                    expected_value = self.expected_result.__getattribute__(name)
+                    if actual_value != expected_value:
+                        log.error(f"{name} invalid: actual {actual_value} expected {expected_value}")
+                        return
+            except AttributeError as error:
+                log.error(f"{error}")
+                return
+        self.expected_result_matched = True
```
Comment (arvidn) [inline tests/plotting/test_plot_manager.py:122]: if you move this in, right before `except`, you don't need the `return` statement here either

**Review by arvidn — DISMISSED** [review]
It's not obvious that this is an improvement. it seem like the objective is to make `expected_result_matched` and `expected_result` not be global variables. To do this, references to them need to be bound to the callback's closure.

You do this by creating a class and make them members of the class and then get a closer of the callback by referencing to `self.refresh_callback`.

In python, variables are implicitly bound to a closures of lambda expressions or local functions. I think that would be a much simpler way of achieving this. Both fewer lines of code and less dependencies of code far apart.

File: tests/plotting/test_plot_manager.py (hunk 2 above)
Comment (xdustinface) [reply tests/plotting/test_plot_manager.py]: Changed

File: tests/plotting/test_plot_manager.py (hunk 1 above)
Comment (xdustinface) [reply tests/plotting/test_plot_manager.py]: Dropped.

**Comment by xdustinface** [comment]
> It's not obvious that this is an improvement. it seem like the objective is to make `expected_result_matched` and `expected_result` not be global variables. 

Correct.

> In python, variables are implicitly bound to a closures of lambda expressions or local functions. I think that would be a much simpler way of achieving this.

If i got you right you mean like in https://github.com/xdustinface/chia-blockchain/commit/00f2ef18ea44407a4d783329fac8d6fc3c0e461f? If so, thats what i had initially but @altendky didn't like it because of the required `nonlocal`, i guess he can elaborate on this. If not, feel free to provide a commit with your idea so that i can pick it in.

**Comment by arvidn** [comment]
> If i got you right you mean like in xdustinface@00f2ef1? If so, thats what i had initially but @altendky didn't like it because of the required nonlocal, i guess he can elaborate on this. If not, feel free to provide a commit with your idea so that i can pick it in.

Ah, yes. That's exactly what I meant. That seems like the more straight-forward way of doing this.

@altendky would you mind elaborating?

**Comment by altendky** [comment]
Discussed on keybase.  The short of it is that I personally find closures to be often misunderstood and also to create references fairly implicitly.  I tend to use them less than I used to, not more.  Certainly they do get used and I'm not about to try to ban them.  `:]`

**Review by arvidn — APPROVED** [review]

**Pull request merged**